extensions/typescript-language-features/src/tsServer/nodeManager.ts

```diff
@@ -0,0 +1,130 @@
+
+import * as vscode from 'vscode';
+import { TypeScriptServiceConfiguration } from '../configuration/configuration';
+import { setImmediate } from '../utils/async';
+import { Disposable } from '../utils/dispose';
+
+
+const useWorkspaceNodeStorageKey = 'typescript.useWorkspaceNode';
+
+export class NodeVersionManager extends Disposable {
+	private _currentVersion: string | undefined;
+
+	public constructor(
+		private configuration: TypeScriptServiceConfiguration,
+		private readonly workspaceState: vscode.Memento
+	) {
+		super();
+
+		this._currentVersion = this.configuration.globalNodePath || undefined;
+		if (vscode.workspace.isTrusted) {
+			if (this.configuration.localNodePath) {
+				if (this.useWorkspaceNodeSetting === undefined) {
+					setImmediate(() => {
+						this.promptAndSetWorkspaceNode();
+					});
+				}
+				else if (this.useWorkspaceNodeSetting) {
+					this._currentVersion = this.configuration.localNodePath;
+				}
+			}
+		}
+		else {
+			this._disposables.push(vscode.workspace.onDidGrantWorkspaceTrust(() => {
+				if (this.configuration.localNodePath) {
+					if (this.useWorkspaceNodeSetting === undefined) {
+						setImmediate(() => {
+							this.promptAndSetWorkspaceNode();
+						});
+					}
+					else if (this.useWorkspaceNodeSetting) {
+						this.updateActiveVersion(this.configuration.localNodePath);
+					}
+				}
+			}));
+		}
+	}
+
+	private readonly _onDidPickNewVersion = this._register(new vscode.EventEmitter<void>());
+	public readonly onDidPickNewVersion = this._onDidPickNewVersion.event;
+
+	public get currentVersion(): string | undefined {
+		return this._currentVersion;
+	}
+
+	public reset(): void {
+		this._currentVersion = undefined;
+	}
+
+	public async updateConfiguration(nextConfiguration: TypeScriptServiceConfiguration) {
+		const oldConfiguration = this.configuration;
+		this.configuration = nextConfiguration;
+		if (oldConfiguration.globalNodePath !== nextConfiguration.globalNodePath
+			|| oldConfiguration.localNodePath !== nextConfiguration.localNodePath) {
+			await this.computeNewVersion();
+		}
+	}
+
+	private async computeNewVersion() {
+		let version = this.configuration.globalNodePath || undefined;
+		if (vscode.workspace.isTrusted && this.configuration.localNodePath) {
+			if (this.useWorkspaceNodeSetting === undefined) {
+				version = await this.promptUseWorkspaceNode();
+			}
+			else if (this.useWorkspaceNodeSetting) {
+				version = this.configuration.localNodePath;
+			}
+		}
+		this.updateActiveVersion(version);
+	}
+
+	private async promptUseWorkspaceNode(): Promise<string | undefined> {
+		const workspaceVersion = this.configuration.localNodePath;
+
+		if (workspaceVersion === null) {
+			throw new Error('Could not prompt to use workspace Node installation because no workspace Node installation is specified');
+		}
+
+		const allow = vscode.l10n.t("Allow");
+		const dismiss = vscode.l10n.t("Dismiss");
+		const neverAllow = vscode.l10n.t("Never in this workspace");
```

Comment:
Once the user selects `never`, is there a way for them still opt in later on?

Reply:
I updated the code to remember the prompt answer for a specific node path. So if you change the workspace node path to point to a different path, the prompt will show up again. However, it's a bit contrived to do that if you select "never" by accident. Should I add a command to use the workspace node path?